### PR TITLE
[jeelink] Add support for emt7110 energy meter

### PR DESCRIPTION
**Pull request by tischober**
# [Jeelink] support for emt7110 energy meter

# Description
This is an update for the jeelink add-on. It support now the emt7110 energy meter and will show the current-power (W), consumptionTotal (kWh), electricPotential (mA) and the electricCurrent (V).

### CODE REVIEW

File: bundles/org.openhab.binding.jeelink/src/main/resources/OH-INF/i18n/jeelink_de.properties [hunk 1]
```diff
@@ -19,6 +19,8 @@ thing-type.ec3k.label = ec3k
 thing-type.ec3k.description = Ein mit dem JeeLink USB Empfänger verbundenes EC3000 Energiekosten-Messgerät.
 thing-type.pca301.label = PCA301
 thing-type.pca301.description = Eine mit dem JeeLink USB Empfänger verbundene PCA301 Steckdose.
+thing-type.emt7110.label = EMT7110
+thing-type.emt7110.description = Eine mit dem JeeLink USB Empfänger verbundene EMT7110 Steckdose.
```
Comment (lsiepel) [inline bundles/org.openhab.binding.jeelink/src/main/resources/OH-INF/i18n/jeelink_de.properties:23]: Please revert changes as only the default language is supplied by github, other languages are done by crowdin.

File: bundles/org.openhab.binding.jeelink/src/main/resources/OH-INF/thing/thing-types.xml [hunk 2]
```diff
@@ -534,4 +566,11 @@
 		<label>@text/channel-type.power-frequency.label</label>
 		<description>@text/channel-type.power-frequency.description</description>
 	</channel-type>
+
+	<channel-type id="voltage-current" advanced="true">
+		<item-type>Number:ElectricPotential</item-type>
+		<label>Voltage</label>
+		<description>Voltage</description>
+		<state readOnly="true" pattern="%.2f %unit%"/>
+	</channel-type>
```
Comment (lsiepel) [inline bundles/org.openhab.binding.jeelink/src/main/resources/OH-INF/thing/thing-types.xml:575]: This channel is added, but it does not seem to be used by the new Thing. I guess this should be removed as the thing allready has two (both potential and current) related and existing channels

**Review by lsiepel — CHANGES_REQUESTED** [review]
Thanks, left some comments to fix.

File: bundles/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/internal/emt7110/Emt7110ReadingConverter.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        }
+
+        return null;
+    }
+    /*
+     * public static void main(String[] args){
+     * Emt7110ReadingConverter emt = new Emt7110ReadingConverter();
+     * emt.createReading("OK EMT7110 84 81 9 91 0 72 0 2 1 6 1");
+     * }
+     */
```
[the remark this answers is not in the document]
Comment (lsiepel) [reply bundles/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/internal/emt7110/Emt7110ReadingConverter.java:56]: This is not resolved yet.

File: bundles/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/internal/emt7110/Emt7110SensorHandler.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+                }
+            }
+
+            @Override
+            public void dispose() {
+            }
+        };
+    }
+
+    private synchronized void cancelRetry() {
```
Comment (lsiepel) [inline bundles/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/internal/emt7110/Emt7110SensorHandler.java:95]: This method partly resolved, it is still synchronized, and i think it should not.

File: bundles/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/internal/emt7110/Emt7110ReadingConverter.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (tischober) [reply bundles/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/internal/emt7110/Emt7110ReadingConverter.java:56]: sorry, missed that...

File: bundles/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/internal/emt7110/Emt7110SensorHandler.java (hunk 4 above)
Comment (tischober) [reply bundles/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/internal/emt7110/Emt7110SensorHandler.java]: i removed the "synchronized".  If that is not right, then I need more informations :-)

File: bundles/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/internal/emt7110/Emt7110SensorHandler.java [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+ *
+ * @author Timo Schober - Initial contribution
+ */
+public class Emt7110SensorHandler extends JeeLinkSensorHandler<Emt7110Reading> {
+    private final Logger logger = LoggerFactory.getLogger(Emt7110SensorHandler.class);
+    private OnOffType state;
+    private JeeLinkHandler bridge;
+    private final AtomicInteger channel = new AtomicInteger(-1);
+
+    private ScheduledFuture<?> retry;
```
Comment (lsiepel) [inline bundles/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/internal/emt7110/Emt7110SensorHandler.java:43]: Looking at this again, what is the purpose none of these seem to be used.

File: bundles/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/internal/emt7110/Emt7110SensorHandler.java [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+    public Class<Emt7110Reading> getReadingClass() {
+        return Emt7110Reading.class;
+    }
+
+    @Override
+    public void initialize() {
+        super.initialize();
+
+        bridge = (JeeLinkHandler) getBridge().getHandler();
+
```
Comment (lsiepel) [inline bundles/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/internal/emt7110/Emt7110SensorHandler.java:59]:
```suggestion
```

File: bundles/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/internal/emt7110/Emt7110SensorHandler.java [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+        logger.debug("initilized handler for thing {} ({})}", getThing().getLabel(), getThing().getUID().getId());
+    }
+
+    @Override
+    public void dispose() {
+        super.dispose();
+        cancelRetry();
```
Comment (lsiepel) [inline bundles/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/internal/emt7110/Emt7110SensorHandler.java:66]:
```suggestion
```

File: bundles/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/internal/emt7110/Emt7110SensorHandler.java [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+        };
+    }
+
+    private void cancelRetry() {
+        ScheduledFuture<?> retry = this.retry;
+        if (retry != null) {
+            retry.cancel(true);
+            retry = null;
+        }
+    }
```
Comment (lsiepel) [inline bundles/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/internal/emt7110/Emt7110SensorHandler.java:101]:
```suggestion
```

File: bundles/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/internal/emt7110/Emt7110SensorHandler.java [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.jeelink.internal.emt7110;
+
+import static org.openhab.binding.jeelink.internal.JeeLinkBindingConstants.*;
+
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.openhab.binding.jeelink.internal.JeeLinkHandler;
```
Comment (lsiepel) [inline bundles/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/internal/emt7110/Emt7110SensorHandler.java:20]:
```suggestion
```

File: bundles/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/internal/emt7110/Emt7110SensorHandler.java [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+import org.openhab.binding.jeelink.internal.JeeLinkSensorHandler;
+import org.openhab.binding.jeelink.internal.ReadingPublisher;
+import org.openhab.core.library.types.OnOffType;
```
Comment (lsiepel) [inline bundles/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/internal/emt7110/Emt7110SensorHandler.java:23]:
```suggestion
```

File: bundles/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/internal/emt7110/Emt7110ReadingConverter.java [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+/**
+ * Handler for a EMT7110 energy Sensor thing.
+ *
+ * @author Timo Schober - Initial contribution
+ */
+public class Emt7110ReadingConverter implements JeeLinkReadingConverter<Emt7110Reading> {
+    private static final Pattern READING_P = Pattern.compile(
+            "OK EMT7110\\s+([0-9]+)\\s+([0-9]+)\\s+([0-9]+)\\s+([0-9]+)\\s+([0-9]+)\\s+([0-9]+)\\s+([0-9]+)\\s+([0-9]+)\\s+([0-9]+)\\s+([0-9]+)\\s+([0-9]+)");
+
+    private final Logger logger = LoggerFactory.getLogger(Emt7110ReadingConverter.class);
```
Comment (lsiepel) [inline bundles/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/internal/emt7110/Emt7110ReadingConverter.java:31]:
```suggestion
```

**Review by lsiepel — CHANGES_REQUESTED** [review]
Seems there where some SAT errors and not used parts. Can you test this PR after these changes have been committed?

Please also update the readme.

**Comment by tischober** [comment]
added the emt7110 to the readme file and also test everything again with openhab4.
I hope I did not forget anything.

**Review by lsiepel — APPROVED** [review]
Thanks, LGTM

**Pull request merged**